### PR TITLE
Output missing cert or key name in DownloadCerts

### LINKED ISSUE #1507: indicate which certificate or key is missing during DownloadCerts
## Is this a BUG REPORT or FEATURE REQUEST?

BUG REPORT

_Although that kind of depends on one's perspective, as we will soon see_

## Versions

**kubeadm version** (use `kubeadm version`):

`kubeadm version: &version.Info{Major:"1", Minor:"14", GitVersion:"v1.14.1", GitCommit:"b7394102d6ef778017f2ca4046abbaa23b88c290", GitTreeState:"clean", BuildDate:"2019-04-08T17:08:49Z", GoVersion:"go1.12.1", Compiler:"gc", Platform:"linux/amd64"}`

**Environment**:
- **Kubernetes version** (use `kubectl version`): `Server Version: version.Info{Major:"1", Minor:"13", GitVersion:"v1.13.4", GitCommit:"c27b913fddd1a6c480c229191a087698aa92f0b1", GitTreeState:"clean", BuildDate:"2019-02-28T13:30:26Z", GoVersion:"go1.11.5", Compiler:"gc", Platform:"linux/amd64"}`
- **Cloud provider or hardware configuration**: `AWS`
- **OS** (e.g. from /etc/os-release): `NAME="Ubuntu" VERSION="18.04.2 LTS (Bionic Beaver)"`
- **Kernel** (e.g. `uname -a`): `Linux ip-10-0-2-238.us-west-2.compute.internal 4.15.0-1035-aws #37-Ubuntu SMP Mon Mar 18 16:15:14 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux`
- **Others**:

## What happened?

While attempting to join a new master, `kubeadm` emitted this error message and halted the join:

`error execution phase control-plane-prepare/download-certs: error downloading certs: couldn't find required certificate or key in Secret`

## What you expected to happen?

`error execution phase control-plane-prepare/download-certs: error downloading certs: couldn't find required certificate or key in Secret[name=external-etcd.crt][path=/etc/kubernetes/pki/apiserver-etcd-client.crt]`

(I patched `kubeadm` and rebuilt a local version to include that information, which is why I was able to finally get the new master to join):

```diff
diff --git a/cmd/kubeadm/app/phases/copycerts/copycerts.go b/cmd/kubeadm/app/phases/copycerts/copycerts.go
index 438a152dae..5c612c6ed8 100644
--- a/cmd/kubeadm/app/phases/copycerts/copycerts.go
+++ b/cmd/kubeadm/app/phases/copycerts/copycerts.go
@@ -231,7 +231,7 @@ func DownloadCerts(client clientset.Interface, cfg *kubeadmapi.InitConfiguration
        for certOrKeyName, certOrKeyPath := range certsToTransfer(cfg) {
                certOrKeyData, found := secretData[certOrKeyNameToSecretName(certOrKeyName)]
                if !found {
-                       return errors.New("couldn't find required certificate or key in Secret")
+                       return fmt.Errorf("couldn't find required certificate or key in Secret[name=%s][path=%s]", certOrKeyName, certOrKeyPath)
                }
                if len(certOrKeyData) == 0 {
                        klog.V(1).Infof("[download-certs] Not saving %q to disk, since it is empty in the %q Secret\n", certOrKeyName, kubeadmconstants.KubeadmCertsSecret)
```

This section of the issue is why this could be a "bug report" or "feature request," or even "close: wontfix" depending on one's perspective.

If it is not a security risk to emit the the `certOrKeyName` in the error message, then I think it's a bug for `kubeadm` to withhold the specifics from the user. Even if it _is_ a security risk, then perhaps something like `--insecure-copycerts-error-message` would allow truly stuck users to opt in to this reduction in security.

## How to reproduce it (as minimally and precisely as possible)?

On an existing master node:

```sh
# /usr/local/bin/kubeadm-v1.14.1 init phase upload-certs --experimental-upload-certs
```

Then on a new master node:

```sh
# /usr/local/bin/kubeadm-v1.14.1 join ${apiserver_name}:443 --token "$KUBEADM_JOIN_TOKEN"  --discovery-token-ca-cert-hash sha256:dfc4b5ada07a6b6a1ee29838eff86a89e36d20ad4521713cb029895bfff70c8e --experimental-control-plane --certificate-key "$KUBEADM_CERT_KEY"
```
and observe the opaque error message because `kubeadm` was looking for the external ETCD certs and not finding them.

Had `kubeadm` chosen to say _which_ key it failed to find, it would have focused my debugging efforts around the missing `--config` to the `upload-certs` phase, rather than trying to guess why `kubeadm` couldn't decrypt the certs.

### PR DESCRIPTION
**Pull request by mdaniel**
Otherwise the user does not know what key `kubeadm` is expecting, which makes troubleshooting harder. See the writeup in 1507 for the reasoning behind this change.

**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:

It improves the error message when `kubeadm` is searching for a key in the `kubeadm-certs` `Secret` but does not find it.

**Which issue(s) this PR fixes**:

Fixes kubernetes/kubeadm#1507

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
Kubeadm will now include the missing certificate key if it is unable to find an expected key during `kubeadm join` when used with the `--experimental-control-plane` flow
```

### CODE REVIEW

File: cmd/kubeadm/app/phases/copycerts/copycerts.go
Comment (yagonobre) [inline cmd/kubeadm/app/phases/copycerts/copycerts.go]: Can you put a space between `Secret` and `[name=%s]`?

**Review by yagonobre — COMMENTED** [review]
Thanks @mdaniel 
Can you sign the CLA?
/ok-to-test
/priority important-soon

**Review by fabriziopandini — APPROVED** [review]
@mdaniel thanks! Welcome!
I'm always happy to see first time contributor choosing kubeadm for their work. We looking forward for more of your contributions; please consider also to join the kubeadm office hours meetings.
/approve
I left final lgtm to @yagonobre / other reviewers

**Comment by yagonobre** [comment]
@mdaniel please squash your commits and we can get it merged :)

**Comment by mdaniel** [comment]
@fabriziopandini thank you for shepherding this change through so quickly; that's in stark contrast to a lot of my other open source experiences. I am sincerely grateful for your help.

> what do you think about print the cert on certOrKeyPath/certOrKeyName format? So we can have something like /etc/kubernetes/pki/ca.crt on logs.

I am a big fan of attaching labels to the text, rather than an implied formatting, but if you feel that another format would be more in alignment with what users would expect, or there is precedent for doing that elsewhere in `kubeadm`, I would defer to the precedent

**Comment by neolit123** [comment]
@yagonobre

> error execution phase control-plane-prepare/download-certs: error downloading certs: couldn't find required certificate or key in Secret[name=external-etcd.crt][path=/etc/kubernetes/pki/apiserver-etcd-client.crt]

the path already includes the file name.
but let's change the formatting a little to be more consistent with our existing errors.

@mdaniel could you please change this to:
```
return errors.Errorf("the Secret does not include the required certificate or key - name: %s, path: %s", certOrKeyName, certOrKeyPath)
```

/hold

**Comment by neolit123** [comment]
thanks for the update.
/lgtm
/hold cancel

**Pull request merged**
